Add missing virtual destructors

ivanpauno opened the pull request:
Recently, we detected a missing virtual dtor in [QOSEventHandlerBase](https://github.com/ros2/rclcpp/pull/1119#discussion_r428216447).
It sounded like something that the compiler could catch, and after looking a bit I found:

>        -Wnon-virtual-dtor (C++ and Objective-C++ only)
>           Warn when a class has virtual functions and an accessible non-virtual destructor itself or in > an accessible polymorphic base class, in which case it is possible but unsafe to delete an instance
>           of a derived class through a pointer to the class itself or base class.  This warning is automatically enabled if -Weffc++ is specified.

i.e.: if the class has a public non virtual dtor and virtual methods, the compiler will warn.

The new compiler option detected missing virtual dtors in `RateBase` and `TimerBase`, which are fixed in this PR too.

---

I propose also adding `-Woveloaded-virtual`:

>        -Woverloaded-virtual (C++ and Objective-C++ only)
>           Warn when a function declaration hides virtual functions from a base class.

We should probably add both warnings to all cpp packages.

---

I'm planning to rebase and merge, one commit adding the warnings and the other fixing the missing virtual dtors.

hidmic commented on rclcpp/CMakeLists.txt:
```diff
@@ -23,7 +23,7 @@ if(NOT CMAKE_CXX_STANDARD)
 endif()
 if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-  add_compile_options(-Wall -Wextra -Wpedantic)
+  add_compile_options(-Wall -Wextra -Wpedantic -Wnon-virtual-dtor -Woverloaded-virtual)
```
@ivanpauno will `-Woverloaded-virtual` result in a warning if said method is both overriden and overloaded?

ivanpauno replied:
gcc man page explanation is better than mine:

> Warn when a function declaration hides virtual functions from a base class

ivanpauno commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10949)](http://ci.ros2.org/job/ci_linux/10949/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6294)](http://ci.ros2.org/job/ci_linux-aarch64/6294/)
* macOS [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8916)](https://ci.ros2.org/job/ci_osx/8916/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10856)](http://ci.ros2.org/job/ci_windows/10856/)

ivanpauno commented:
@jacobperron @hidmic I've added the same flags to `rclcpp_action`, `rclcpp_components` and `rclcpp_lifecycle`; added some missing virtual dtors in `rclcpp_lifecycle` and retriggered CI.

ivanpauno commented:
CI failed because `LifecyclePublisherInterface` destructor didn't have a definition (IDK what I was thinking ...). It didn't show up until `lifecycle` demo was built, because we don't have a test for `LifecyclePublisherInterface` in `rclcpp_lifecycle`.

Retriggered full CI in https://github.com/ros2/rclcpp/pull/1149#issuecomment-636981657...

ivanpauno commented:
@ros-pull-request-builder retest this please

wjwwood commented:
lgtm since build.ros2.org passed the linter fix.
